src/extension/app-schema/app-schema-web/src/main/java/org/geoserver/complex/FeatureTypeConverter.java

```diff
@@ -0,0 +1,187 @@
+/* (c) 2020 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.complex;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.logging.Logger;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
+import org.geotools.filter.AttributeExpressionImpl;
+import org.geotools.util.logging.Logging;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.feature.type.AttributeDescriptor;
+import org.opengis.feature.type.AttributeType;
+import org.opengis.feature.type.FeatureType;
+import org.opengis.feature.type.GeometryDescriptor;
+import org.opengis.feature.type.Name;
+import org.opengis.feature.type.PropertyDescriptor;
+import org.xml.sax.helpers.NamespaceSupport;
+
+/**
+ * Converts a complex feature type to a simple one based on the provided transformation rules map.
+ */
+public class FeatureTypeConverter {
```

Comment:
Guess also this class could have been package private. In general, I'd suggest to limit visibility to the minimum actually needed, makes easier to change classes/methods later, no backwards compatibility to preserve.